flang/lib/Semantics/check-directive-structure.h

```diff
@@ -198,6 +198,10 @@ class DirectiveStructureChecker : public virtual BaseChecker {
     ClauseMapTy clauseInfo;
     std::list<C> actualClauses;
     std::list<C> crtGroup;
+    std::set<std::string> usedInScanDirective;
```

Comment:
Hi Sergio, I think it is better to get represented in directive context stack . So once the control goes out of the reduction (directive context), the symbol is no longer marked. Otherwise, some of the semantic checks may be invalid.

Reply:
The `symbol *` of reduction variables and scan variables are different now. Reduction declares a private symbol as implemented [here](https://github.com/anchuraj/llvm-project/blob/master/flang/lib/Semantics/resolve-directives.cpp#L2478). So checking `Symbol *` would not be appropriate at this stage. Looking to make further modification of using the same `Symbol *` in both clauses (if required) as I proceed with the implementation.

Reply:
I agree with @skatrak's initial comments. We should try to keep the OpenMP specific data-structures in the OpenMP specific checkers. And if we can use the symbol that would be great. If the reduction creates a new private symbol, that should be the symbol bound to the  inclusive or exclusive clause of the SCAN directive.

Reply:
We are probably missing a `ResolveName` for the names in the clauses in the scan directive. This could be the reason that we are not getting the same symbols in the reduction clause and the scan directive.

Reply:
Thank you. Missing `ResolveName` was the reason. I have updated. I was able to update the datastructures to use `Symbol *` I have moved the structures to `check-omp-structure.h`